tooling/nargo_cli/src/cli/test_cmd.rs

```diff
@@ -143,27 +190,38 @@ fn run_tests<S: BlackBoxFunctionSolver<FieldElement> + Default>(
     let plural = if count_all == 1 { "" } else { "s" };
     println!("[{}] Running {count_all} test function{plural}", package.name);
 
-    let test_report: Vec<(String, TestStatus)> = test_functions
-        .into_par_iter()
+    let tests: Vec<Test> = test_functions
+        .into_iter()
         .map(|test_name| {
-            let status = run_test::<S>(
-                file_manager,
-                parsed_files,
-                package,
-                &test_name,
-                show_output,
-                foreign_call_resolver_url,
-                root_path.clone(),
-                package_name.clone(),
-                compile_options,
-            );
-
-            (test_name, status)
+            let test_name_copy = test_name.clone();
+            let root_path = root_path.clone();
+            let package_name_clone = package_name.clone();
+            let runner = Box::new(move || {
+                let test_status = run_test::<S>(
+                    file_manager,
+                    parsed_files,
+                    package,
+                    &test_name,
+                    show_output,
+                    foreign_call_resolver_url,
+                    root_path,
+                    package_name_clone.clone(),
+                    compile_options,
+                );
+                display_test_status(
+                    &test_name,
+                    &package_name_clone,
+                    &test_status,
+                    file_manager,
+                    compile_options,
+                );
```

Comment:
Maybe I'm misunderstanding what this will do, but I reckon if you put the display into the parallel bit it might get garbled, which is why it would be good to collect the results into a channel, which a single worker displays one by one.

Reply:
Oh I see there is a lock on the writer. But is it a different instance within each tester?

Reply:
Yes, I don't know if the lock locks the underlying stream or not. I checked the docs and it's not clear.

I'm also thinking we could remove the lock as now all the printing happens sequentially... except that Noir's `println` might get in the middle... but then I guess the writer's lock doesn't make sense, it can't possibly lock the underlying stream.

That said, I tried printing stuff from Noir and the output never got garbled, not sure why. Maybe Rust's println is atomic?

Reply:
I added a test like this and can confirm that the `.lock()` does have the desired effect:
```rust
#[cfg(test)]
mod tests {
    use std::io::Write;
    use std::{thread, time::Duration};
    use termcolor::{ColorChoice, StandardStream};

    #[test]
    fn test_stderr_lock() {
        for i in 0..4 {
            thread::spawn(move || {
                let mut writer = StandardStream::stderr(ColorChoice::Always);
                let mut writer = writer.lock();

                let mut show = |msg| {
                    thread::sleep(Duration::from_millis(10));
                    // println!("{i} {msg}");
                    writeln!(writer, "{i} {msg}").unwrap();
                };

                show("a");
                show("b");
                show("c");
            });
        }
        thread::sleep(Duration::from_millis(100));
    }
}
```

As above it prints:
```
0 a
0 b
0 c
1 a
1 b
1 c
2 a
2 b
2 c
3 a
3 b
3 c
```

If we remove the ` let mut writer = writer.lock();` line it gets garbled:
```
1 3 a
0 a
2 a
a
01 b
3 b
2 b
 b
1 c
3 c
2 c
0 c
```

And if we then use `println!` it gets just mixed (so, no `.lock()`):
```
2 a
0 a
1 a
3 a
2 b
0 b
3 b
1 b
2 c
0 c
3 c
1 c
```

So I'd expect that if we execute tests in parallel then yes, `println!` will not be garbled per-se, but the output of different Noir tests can get intertwined.

The reports at the end will be good, so if the output was collected instead of printed, then the display can happens as you moved it into the tester, rather than the way I suggested with the async collector.

Reply:
> but the output of different Noir tests can get intertwined

What I noticed is that the output of different Rust tests can also get intertwined. That was my motivation for not making it work better in Noir 😅

Reply:
Fair enough :D 
Is that always the case, or only when `--nocapture` is used?

Reply:
I think we should do this in a followup PR as there's the same potential for garbling currently, right?